[FLINK-27101][checkpointing][rest] Add restful API to trigger checkpoints

leletan opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

This pull request adds a restful API for one-time checkpointing triggering. This allows user to break incremental checkpoint chain by triggering a manual full checkpoint. Restful API delegates the scheduling to user.

## Brief change log

  - The checkpoint trigger restful API is async, similar to savepoint restful API
  - The checkpoint trigger API takes a CheckpointBackupType and will only trigger checkpoint once
  - A new public triggerCheckpoint method taking a CheckpointBackupType is added to CheckpointCoordinator

## Verifying this change

This change added tests and can be verified as follows:
  - Added unit tests to validate CheckpointTriggerHandler and CheckpointTriggerStatusHandler restful handlers

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: yes
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? generated docs for restful API

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointRetentionPolicy.java:
```diff
@@ -24,6 +24,9 @@
 public enum CheckpointRetentionPolicy {
 
+    /** Full Checkpoints should be retained on cancellation and failure. */
+    FULL_RETAIN_ON_CANCELLATION,
```
Whether the checkpoint is full or not doesn't fit with the `CheckpointRetentionPolicy`. It should be an independent axis.

If the checkpoint is full or incremental it should be parametrised independently of the retention policy. Assuming that we want to expose retention to users - and I don't see why we should be doing this. Is there some particular reason why have you implemented it like this @leletan ? Such exposure makes this more complicated for us to maintain code in the future and it also makes it more complicated for the users. To me it sounds like this retention policy should come from the `CheckpointCoordinatorConfiguration` and be the same for manually and automatically triggered checkpoints.

leletan replied:
Good catch @pnowojski. This was one of the place I was not very confident with. As a hindsight, totally agree it was a bad design.

The reason why I did that was as following:
I was not sure if a CheckpointProperty will be valid if it is too far away from [these](https://github.com/apache/flink/blob/d8a4304b892412eab4c5c19b5deb84166943d3bb/flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointProperties.java#L246) pre-configured ones so I was thinking of creating one as close to an existing one as possible. Meanwhile the public method to create a pre-configured CheckpointProperty was through [this](https://github.com/apache/flink/blob/d8a4304b892412eab4c5c19b5deb84166943d3bb/flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointProperties.java#L349) method with a `CheckpointRetentionPolicy`.

Meanwhile I would love to hear your thoughts 
1. What checkpoint configurations we should expose in the restful API? Are there more general use cases requesting other configurations for checkpoint manual triggering (other than ours)?
2. Or would it be better just to start with our specific use case - always only trigger a full checkpoint, i.e.  reuse all `CheckpointProperty` and only replace `CheckpointType` with [`FULL_CHECKPOINT `](https://github.com/apache/flink/blob/d8a4304b892412eab4c5c19b5deb84166943d3bb/flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointType.java#L30)

pnowojski replied:
I could see use cases for manually triggering incremental checkpoints as well, so I would suggest to full/incremental parameter. We would need to decouple REST API constants from the internal classes and add a translation layer for that. It should be probably something similar to `org.apache.flink.core.execution.SavepointFormatType`. `CheckpointType` with two supported values `INCREMENTAL` and `FULL`. 

As for the `CheckpointProperties` class, probably you should copy instead of passing `CheckpointProperties` in `org.apache.flink.runtime.checkpoint.CheckpointCoordinator#triggerCheckpoint(org.apache.flink.runtime.checkpoint.CheckpointProperties)`, you should just pass `org.apache.flink.runtime.checkpoint.CheckpointType` and inside the `triggerCheckpoint` construct new `CheckpointProperties` by copying `org.apache.flink.runtime.checkpoint.CheckpointCoordinator#checkpointProperties`. Something like that:
```

    public CompletableFuture<CompletedCheckpoint> triggerCheckpoint(CheckpointType checkpointType) {
        CheckpointProperties properties =
                new CheckpointProperties(
                        checkpointProperties.forceCheckpoint(),
                        checkpointType,
                        checkpointProperties.discardOnSubsumed(),
                        checkpointProperties.discardOnJobFinished(),
                        checkpointProperties.discardOnJobCancelled(),
                        checkpointProperties.discardOnJobFailed(),
                        checkpointProperties.discardOnJobSuspended(),
                        checkpointProperties.isUnclaimed());
        return triggerCheckpointFromCheckpointThread(properties, null, false);
    }
```

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterGateway.java:
```diff
@@ -218,13 +220,27 @@ CompletableFuture<String> triggerSavepoint(
+     * @param timeout for the rpc call
+     * @return Future which is completed with the savepoint path once completed
+     */
+    CompletableFuture<CompletedCheckpoint> triggerCheckpoint(
```
Does it have to return `CompletedCheckpoint`? It makes a bit inconsistent with the other method. Can not we use checkpoint location as the cache key?

leletan replied:
Good question, @pnowojski. 
Assuming we are talking about 
1. `CompletableFuture<String> triggerCheckpoint(@RpcTimeout final Time timeout)` as the other method 
2. `CompletedOperationCache<AsynchronousJobOperationKey, Long> checkpointTriggerCache` in the `DispatcherCachedOperationsHandler` as the cache (location as cache value?)
We definitely can. Meanwhile I guess this means that the restful API (checkpoint trigger status API) will eventually return checkpoint location in the restful response body, instead of checkpoint id.

The reason for returning `CompletedCheckpoint` was to facilitate both old method and new API: the new restful API was returning checkpoint id, while the other method was returning checkpoint location, and CompletedCheckpoint contains both.

I am pretty new to flink so I am not sure what are people typically care about a checkpoint.
If we think it is a better design to return checkpoint location, instead of checkpoint id in the response body in the restful API, I am happy to return checkpoint location and reuse existing cache and another method.

pnowojski replied:
Good point. Indeed it would be better to not expose checkpoint location, but just the checkpoint id. Checkpoints are owned by the system (contrary to the savepoints, which are owned by the user), so user shouldn't be tempted to touch checkpoint files. In this case, let's keep `CompletedCheckpoint`

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterGateway.java:
```diff
@@ -218,13 +220,27 @@ CompletableFuture<String> triggerSavepoint(
+     * @param checkpointProperties to determine how checkpoint should be taken or null if the
+     *     existing checkpoint property should be used
+     * @param timeout for the rpc call
+     * @return Future which is completed with the savepoint path once completed
```
nit: java doc should say it returns `{@link CompletedCheckpoint}` not `path`.

leletan replied:
Good catch

pnowojski reviewed (changes requested):
Thanks a lot for the contribution @leletan . It looks mostly good to me, I've left a couple of comments. It would be great if someone else (maybe @zentol ) could also take a look at the REST API changes.

Also there seems to be a build failure.

leletan commented:
> Thanks a lot for the contribution @leletan . It looks mostly good to me, I've left a couple of comments. It would be great if someone else (maybe @zentol ) could also take a look at the REST API changes.
> 
> Also there seems to be a build failure.
> 
> > If yes, how is the feature documented? [WIP]
> 
> Let's not forget about this :)

Thanks for the super fast response, @pnowojski !

Do we have a standard convention/process about restful API feature? 
- The only thing I found so far is [this](https://github.com/apache/flink/tree/master/flink-docs#rest-api-documentation) README. I am still a little bit confused after reading the README. Do we have an example for that?
- Any restful API documentation other than above we need to add?

leletan commented:
Also I am wondering if this change would need any further test coverages? If yes, where would be the best place I should start looking?

zentol commented:
re documentation: You basically just run `mvn package -Dgenerate-rest-docs -pl flink-docs -am -nsu -DskipTests` locally.

leletan commented:
@pnowojski @zentol Addressed the code comments (at least I thought I did) and generated the docs. Please take a look whenever you have time. Thanks!

pnowojski reviewed (commented):
Thanks for the update. I've left one more comment.

I see one unrelated test failure because of FLINK-29315 and another one about REST API compatibility. Has it failed because there was some accidental braking change or just because of adding a new REST handler?

Re the test coverage. From the perspective of the `CheckpointCoordinator`. It would be nice to check that indeed full or incremental checkpoints are triggered as requested. Maybe you could add some test similar to `org.apache.flink.runtime.checkpoint.CheckpointCoordinatorTriggeringTest#testTriggeringFullCheckpoints`?  But instead of using `manuallyTriggeredScheduledExecutor`, to manually trigger periodic checkpoints, just simply use directly the newly added `CheckpointCoordinator#triggerCheckpoint` method? Re the REST API tests, let's see what  @zentol will have to say.

zentol commented:
`API was modified in a compatible way, but the snapshot was not updated. To update the snapshot, re-run this test with -Dgenerate-rest-snapshot being set.`

Run "mvn verify -Dgenerate-rest-snapshot -pl flink-runtime-web -nsu", or `RuntimeRestAPIStabilityTest` with the property being set.

Essentially there are 2 files we generate: the documentation (with my last comment) and a snapshot for comparing compatibility (we cant use the docs because we don't persist everything we need IIRC).

stevenzwu commented on flink-core/src/main/java/org/apache/flink/core/execution/CheckpointBackupType.java:
```diff
@@ -0,0 +1,52 @@
+
+/** Describes the backup type in which a checkpoint should be taken. */
+@PublicEvolving
+public enum CheckpointBackupType implements DescribedEnum {
```
@leletan We probably don't need this enum at all and existing `CheckpointType` should we sufficient. There can be two triggering modes for checkpoint via REST endpoint.

- `CHECKPOINT`: checkpoint is triggered based on the job configured checkpoint mode (incremental or full)
- `FULL_CHECKPOINT`: checkpoint is triggered for full checkpoint regardless of the job config

If in the future there is a need for triggering incremental checkpoint regardless of the job config, we can add a new enum value `INCREMENTAL_CHECKPOINT` to the existing `CheckpointType` later.

pnowojski replied:
Already existing `org.apache.flink.runtime.checkpoint.CheckpointType` is an internal class, not a public one and I asked to not expose it the users. Note for example `org.apache.flink.runtime.checkpoint.CheckpointType#getSharingFilesStrategy`, which shouldn't be exposed. 

https://github.com/apache/flink/pull/20852#discussion_r975199952

Hence I suggested to create this public facing one.

stevenzwu replied:
Thanks for the context @pnowojski . then we need to add the `@Internal` annotation to the `org.apache.flink.runtime.checkpoint.CheckpointType` class. I didn't see the annotation and hence thought it is intended as public.

With that context, I would agree with Piotr's suggestion on renaming the internal class.  maybe we can rename `org.apache.flink.runtime.checkpoint.CheckpointType` as `org.apache.flink.runtime.checkpoint.SnapshotTypeImpl` and annotate it with `@Internal`.

pnowojski replied:
`SnapshotTypeImpl` is not a great name choice, as the there is another implementation of `SnapshotType`.

Maybe let's rename the old `CheckpointType` and `SavepointType` to  `CheckpointInternalType` and `SavepointInternalType`? 🤷

stevenzwu replied:
agree that `SnapshotTypeImpl` is not a good choice. `Internal` would be ok for me. 

Just to brainstorm more options. what about `CheckpointSnapshotType` and `SavepointSnapshotType`?

pnowojski replied:
> `CheckpointSnapshotType` and `SavepointSnapshotType`

Either this or `Internal` infix are fine by me 👍

leletan replied:
Renamed `CheckpointType` -> `CheckpointSnapshotType` and `SavepointType` -> `SavepointSnapshotType`

stevenzwu commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -511,6 +514,45 @@ public CompletableFuture<CompletedCheckpoint> triggerCheckpoint(boolean isPeriod
+            @Nullable CheckpointBackupType checkpointBackupType) {
+
+        if (checkpointBackupType == null) {
+            return triggerCheckpointFromCheckpointThread(checkpointProperties, null, false);
```
might be a little cleaner to avoid calling `trigger...` here. we can just resolve the `checkpointType` and call the `trigger...` at line 543. Also if we directly use `CheckpointType` and removed the `CheckpointBackupType`, this can be even simpler.

stevenzwu commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherCachedOperationsHandler.java:
```diff
@@ -39,30 +40,66 @@ public class DispatcherCachedOperationsHandler {
+        return CompletableFuture.completedFuture(Acknowledge.get());
+    }
+
+    public CompletableFuture<OperationResult<Long>> getCheckpointStatus(
```
The result type is `Long`. what does the long value result indicate? I assume the `Long` type is not meant for checkpoint Id.

stevenzwu replied:
I got it now after seeing the other discussion. it is for checkpoint id. For savepoint trigger, the result is a path/location. for checkpoint trigger, the result is a checkpoint id.

stevenzwu commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMaster.java:
```diff
@@ -863,11 +871,6 @@ public CompletableFuture<String> triggerSavepoint(
         return schedulerNG.triggerSavepoint(targetDirectory, cancelJob, formatType);
     }
 
-    @Override
```
we shouldn't remove a public method from a public class. it is also necessary here. it can be kept for backward compatibility.

leletan replied:
The backward compatibility should already been kept. At this class's parent interface `JobMasterGateway`, this method is supported as 

```
    /**
     * Triggers taking a checkpoint of the executed job.
     *
     * @param timeout for the rpc call
     * @return Future which is completed with the checkpoint path once completed
     */
    default CompletableFuture<String> triggerCheckpoint(@RpcTimeout final Time timeout) {
        return triggerCheckpoint(null, timeout).thenApply(CompletedCheckpoint::getExternalPointer);
    }

stevenzwu replied:
got it. you added a default impl method in the parent interface.

stevenzwu commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMaster.java:
```diff
@@ -853,6 +855,12 @@ public CompletableFuture<ExecutionGraphInfo> requestJob(Time timeout) {
     }
 
+    @Override
+    public CompletableFuture<CompletedCheckpoint> triggerCheckpoint(
```
nit: this can be added next to the other `triggerCheckpoint` method

leletan replied:
There will be no other `triggerCheckpoint` method in this class, that method is implemented in its parent interface so it will not be defined explicitly here.

stevenzwu commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/checkpoints/CheckpointTriggerHandlers.java:
```diff
@@ -0,0 +1,261 @@
+ * <p>Drawing checkpoints is a potentially long-running operation. To avoid blocking HTTP
+ * connections, checkpoints must be drawn in two steps. First, an HTTP request is issued to trigger
+ * the checkpoint asynchronously. The request will be assigned a {@link TriggerId}, which is
+ * returned in the response body. Next, the returned id should be used to poll the status of the
```
nit: `returned id` -> `returned {@link TriggerId}`

leletan commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:
```diff
@@ -878,7 +879,8 @@ public CompletableFuture<String> triggerSavepoint(
 
     @Override
-    public CompletableFuture<String> triggerCheckpoint() {
+    public CompletableFuture<CompletedCheckpoint> triggerCheckpoint(
```
Hmm, good question ... I was just following the same function call path as triggerSavepoint (resful -> dispatcher -> job master -> scheduler -> checkpoint coordinator). 
@stevenzwu Are you suggesting we skip scheduler and call CheckpointCoordinator::triggerCheckpoint(...) directly from job master?

stevenzwu replied:
oh. I didn't realize it is needed in the trigger checkpoint code path.

leletan commented:
Hi folks, @pnowojski @stevenzwu @zentol I believe I addressed all comments except the pending naming of the old CheckpointType class name. 
Personally I don't have any strong opinion about the class name so I just marked it as `@Internal` for now. I am happy to update it after we reach a consensus. 
Meanwhile please let me know if everything else looks good to you. 
Thanks!

leletan commented:
Thanks again folks. All comments should be address, please let me know if there is anything else I need to do for this PR.

stevenzwu commented on flink-core/src/main/java/org/apache/flink/core/execution/CheckpointType.java:
```diff
@@ -0,0 +1,52 @@
+@PublicEvolving
+public enum CheckpointType implements DescribedEnum {
+    /** A checkpoint type with the entire state check pointed, common for all state backends. */
+    FULL("A checkpoint type with the entire state check pointed, common for all state backends."),
```
should be one world: check pointed -> checkpointed

stevenzwu commented on flink-core/src/main/java/org/apache/flink/core/execution/CheckpointType.java:
```diff
@@ -0,0 +1,52 @@
+    FULL("A checkpoint type with the entire state check pointed, common for all state backends."),
+
+    /**
+     * A checkpoint type that check points only the difference between snapshots, specific for
```
should be one word: check points -> checkpoints. 

Also I see the style difference in the description btw full and incremental. We probably can reword the description for full as following.
```
A checkpoint type that checkpoints the entire state, ...
```

leletan replied:
This is copied from [SavepointFormatType](https://github.com/apache/flink/blob/f543b8ac690b1dee58bc3cb345a1c8ad0db0941e/flink-core/src/main/java/org/apache/flink/core/execution/SavepointFormatType.java)
But I am fine with removing it.

stevenzwu commented on flink-core/src/main/java/org/apache/flink/core/execution/CheckpointType.java:
```diff
@@ -0,0 +1,52 @@
+     * certain state backend.
+     */
+    INCREMENTAL(
+            "A checkpoint type that check points only the difference between snapshots, specific for certain state backend.");
```
should be one word: check points -> checkpoints. 

Also I see the style difference in the description btw full and incremental. We probably can reword the description for full as following.
```
A checkpoint type that checkpoints the entire state, ...
```

stevenzwu commented on flink-core/src/main/java/org/apache/flink/core/execution/CheckpointType.java:
```diff
@@ -0,0 +1,52 @@
+/** Describes the type in which a checkpoint should be taken. */
+@PublicEvolving
+public enum CheckpointType implements DescribedEnum {
+    /** A checkpoint type with the entire state check pointed, common for all state backends. */
```
the Javadoc seems unnecessary considering the description already in place.

leletan replied:
This is copied from [SavepointFormatType](https://github.com/apache/flink/blob/f543b8ac690b1dee58bc3cb345a1c8ad0db0941e/flink-core/src/main/java/org/apache/flink/core/execution/SavepointFormatType.java)
But I am fine with removing it.

stevenzwu commented on flink-core/src/main/java/org/apache/flink/core/execution/CheckpointType.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    @Override
+    @Internal
```
this annotation is not necessary, right?

leletan replied:
This is copied from [SavepointFormatType](https://github.com/apache/flink/blob/f543b8ac690b1dee58bc3cb345a1c8ad0db0941e/flink-core/src/main/java/org/apache/flink/core/execution/SavepointFormatType.java#L50)
But I am fine with removing it.

stevenzwu replied:
I think we should remove it. a little weird to annotate a override method for a public interface.

stevenzwu commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -511,6 +514,42 @@ public CompletableFuture<CompletedCheckpoint> triggerCheckpoint(boolean isPeriod
+
+        final SnapshotType snapshotType;
+        if (checkpointType == null) {
+            snapshotType = checkpointProperties.getCheckpointType();
```
we should rename the `getCheckpointType` method too

stevenzwu commented:
@leletan I think we probably should make the `CheckpointType` and `SavepointType` renaming as a separate PR. merge that one first then come back on this one. it gets difficult to review the diff when the two changes are mixed together. touching so many files.

stevenzwu reviewed (changes requested):
please move the `SnapshotType` impl class renaming to a separate JIRA and PR.

leletan commented:
@stevenzwu All your comments should be addressed. Please let me know if there is anything missing. Thanks!

leletan commented:
@pnowojski Your comments and change requests should all be addressed. Please let me know if I miss anything. Thanks!

stevenzwu commented:
@leletan we shall come back to this PR after the internal class renaming is merged. Please create a jira and PR for that first.

pnowojski reviewed (approved):
> @leletan I thought we will come back to this PR after the internal class renaming is merged, unless @pnowojski thinks we can proceed with this first. Please create a jira for the renaming.

I'm actually fine either way. Definitely rename should be a separate commit. I think Ideally rename should be done in this PR, with the same JIRA ticket number, as a first, but indeed a separate commit. But it's ok for me if it's an independent JIRA/pr merged after or before this one.

Apart of that +1 @leletan . Thanks for the contribution, and thanks @stevenzwu  for help with reviewing this one :)

leletan commented:
Thanks, @pnowojski and @stevenzwu! 
Combing the feedbacks and here is my plan:
1. fix the build
2. rebase all the changes so far into one commit. 
3. do a follow up commit for the class renaming in this PR
Let me know if you have concern about above.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointTriggerInfo.java:
```diff
@@ -0,0 +1,78 @@
+
+    private static final String FIELD_NAME_CHECKPOINT_ID = "checkpointId";
+
+    private static final String FIELD_NAME_FAILURE_CAUSE = "failure-cause";
```
```suggestion
    private static final String FIELD_NAME_FAILURE_CAUSE = "failureCause";
```

leletan replied:
Was thinking of the same, but hesitated when seeing this in the [SavepointInfo](https://github.com/apache/flink/blob/99c2a415e9eeefafacf70762b6f54070f7911ceb/flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/savepoints/SavepointInfo.java#L42), should we fix that as well?

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointTriggerMessageParameters.java:
```diff
@@ -0,0 +1,43 @@
+/** The parameters for triggering a checkpoint. */
+public class CheckpointTriggerMessageParameters extends MessageParameters {
+
+    public JobIDPathParameter jobID = new JobIDPathParameter();
```
```suggestion
    public final JobIDPathParameter jobID = new JobIDPathParameter();
```

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointTriggerInfo.java:
```diff
@@ -0,0 +1,78 @@
+        this.failureCause = failureCause;
+    }
+
+    @Nullable
+    public Long getCheckpointId() {
+        return checkpointId;
+    }
+
+    @Nullable
+    public SerializedThrowable getFailureCause() {
+        return failureCause;
+    }
```
add `@JsonIgnore`

leletan replied:
Sure. Do we want to fix [SavepointInfo](https://github.com/apache/flink/blob/99c2a415e9eeefafacf70762b6f54070f7911ceb/flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/job/savepoints/SavepointInfo.java#L74) as well?

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherOperationCaches.java:
```diff
@@ -40,13 +42,19 @@ public DispatcherOperationCaches() {
+
     public CompletableFuture<Void> shutdownCaches() {
-        return savepointTriggerCache.closeAsync();
+        return CompletableFuture.allOf(
```
FutureUtils#completeAll is preferably because it doesn't drop exceptions should multiple futures fail.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointTriggerHeaders.java:
```diff
@@ -0,0 +1,77 @@
+        return URL;
+    }
+
+    public static CheckpointTriggerHeaders getInstance() {
```
strange that this is placed in.betweeen getters.

leletan replied:
will fix

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointTriggerStatusHeaders.java:
```diff
@@ -0,0 +1,82 @@
+        return URL;
+    }
+
+    public static CheckpointTriggerStatusHeaders getInstance() {
```
same as above; should be at the very bottom

leletan replied:
good catch, will fix

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerNG.java:
```diff
@@ -125,7 +127,8 @@ void notifyKvStateUnregistered(
 
-    CompletableFuture<String> triggerCheckpoint();
+    CompletableFuture<CompletedCheckpoint> triggerCheckpoint(
+            @Nullable CheckpointType checkpointType);
```
Is this actually nullable? Seems strange give that its not nullable in the request body.

zentol replied:
Why cant we have the MiniCluster default to `CheckpointType#DEFAULT`, or whatever was set as the default based on the configuration.

leletan replied:
Good point. Let me see if I can do this without breaking any public API

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/checkpoints/CheckpointTriggerHandlersTest.java:
```diff
@@ -0,0 +1,355 @@
+                        .get()
+                        .getTriggerId();
+
+        AsynchronousOperationResult<CheckpointTriggerInfo> checkpointTriggerResponseBody;
+        checkpointTriggerResponseBody =
```
```suggestion
        final AsynchronousOperationResult<CheckpointTriggerInfo> checkpointTriggerResponseBody =
```

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/checkpoints/CheckpointTriggerHandlersTest.java:
```diff
@@ -0,0 +1,355 @@
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpResponseStatus;
+
+import org.hamcrest.Matchers;
+import org.junit.Before;
+import org.junit.Test;
```
new tests should use junit5 and assertj.

zentol reviewed (changes requested):
Found a few things that need to be fixed, and a rebase is required.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointTriggerRequestBody.java:
```diff
@@ -0,0 +1,63 @@
+public class CheckpointTriggerRequestBody implements RequestBody {
+
+    private static final String FIELD_NAME_TRIGGER_ID = "triggerId";
+    private static final String FIELD_NAME_CHECKPOINT_TYPE = "checkpointType";
```
Why is this different that what is documented in the docs?

leletan replied:
Oops, good catch. 
There were some code changes after generating the doc based on PR comments, forgot to update doc. Will update docs.

leletan commented on flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java:
```diff
@@ -902,6 +905,28 @@ public CompletableFuture<String> triggerCheckpoint(JobID jobID, Time timeout) {
+        return dispatcherCachedOperationsHandler.getCheckpointStatus(operationKey);
+    }
+
+    public CompletableFuture<Long> triggerCheckpointAndGetCheckpointID(
```
Good catch. Was just following the [corresponding savepoint function](https://github.com/apache/flink/blob/9152af41c2d401e5eacddee1bb10d1b6bea6c61a/flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java#L916-L931). Just realized this could be unnecessary since this function is not needed in [EmbeddedJobClient](https://github.com/apache/flink/blob/cd8ea8d5b207569f68acc5a3c8db95cd2ca47ba6/flink-clients/src/main/java/org/apache/flink/client/deployment/application/EmbeddedJobClient.java)

leletan replied:
Will make it private

leletan commented:
@flinkbot run azure

leletan commented:
Hi folks, @pnowojski @stevenzwu @zentol Thanks for your review again. I believe I addressed all comments, except the pending naming of the old CheckpointType class name (will add in a later commit to make the code easier)
One big change is, I added one more CheckpointType as `CONFIGURED`, to represent using the existing internal CheckpointSnapshotType as configured by flink application to triggerCheckpoint, replacing previous legacy minicluster case represented by `null`. 

Please let me know if there is anything else I am missing.

leletan commented on flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/job/checkpoints/CheckpointTriggerHandlersTest.java:
```diff
@@ -0,0 +1,349 @@
+                        .get();
+
+        assertEquals(checkpointTriggerResponseBody.queueStatus().getId(), QueueStatus.Id.COMPLETED);
+        assertNotNull(checkpointTriggerResponseBody.resource());
```
I see what you mean by assertj now. Will fix.

leletan commented:
@zentol I see I have a lot of style issues and thanks for pointing them out. I am wondering if you have a doc / pointer for those style rules so I won't make those mistakes in the future.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointTriggerStatusHeaders.java:
```diff
@@ -0,0 +1,82 @@
+import org.apache.flink.shaded.netty4.io.netty.handler.codec.http.HttpResponseStatus;
+
+/** These headers define the protocol for triggering a checkpoint. */
+public class CheckpointTriggerStatusHeaders
+        extends AsynchronousOperationStatusMessageHeaders<
+                CheckpointTriggerInfo, CheckpointTriggerStatusMessageParameters> {
```
> this is very similar to savepoint logic, which has no "trigger" for its API or class names.

That's why I'd like avoid it. It behaves the same way but currently has a different naming scheme.

> 1. The new APIs in this PR is for manual trigger of a checkpoint and it may or may not end up with a checkpoint (it could end up with a failure and return with a failure cause, before it can reach CheckpointCoordinator)

That's a good observation but I'd say it's fine nonetheless to not refer to "triggers".
It's just a technical limitation that we can't differentiate between failures before/after the checkpoint/savepoint was triggered that we _may_ fix at some point.
I also think this distinction isn't really important. It would of course be great if the initial request could already fail early, but at the end of the day the results are the same to the user.

> 2. The checkpoint trigger status is queried by trigger id and its value is read from DispatcherOperationCaches which has only 5 min cache duration by default, so it is more of a trigger status than checkpoint status.

The timeout is there because the JM waits with it's shutdown until the result of these requests have been consumed, and we don't want it to sit there idling indefinitely because someone forget to consume a checkpoint result.

> 3. There are already other APIs for querying checkpoint metadata and details, which should cover all the retained checkpoints for the application - don't want to confuse user with them.

Interesting point. However this also applies to the savepoints API and so far no one complained. I think there's a well-defined use-cases for both checkpoint-related APIs; one gives you the full details, the other just status information about manually triggered checkpoints (which is somewhat explicit since the url has a triggerId in it).
I wouldn't worry about it.

leletan replied:
I see. Will refactor to match savepoint naming.

leletan commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/checkpoints/CheckpointTriggerHandlers.java:
```diff
@@ -0,0 +1,261 @@
+            return gateway.triggerCheckpoint(
+                            operationKey,
+                            request.getRequestBody().getCheckpointType(),
+                            RpcUtils.INF_TIMEOUT)
```
Created [this](https://issues.apache.org/jira/browse/FLINK-29610) JIRA.

leletan commented:
@zentol Thanks for the detailed feedbacks, really appreciated them!
I should have addressed all the comments in this round. Please take another look when you have a chance.

zentol commented:
@flinkbot run azure
